src/Markup/Perspex.Markup/Data/Parsers/LiteralParser.cs

```diff
@@ -12,21 +12,33 @@ public static object Parse(Reader r)
         {
             if (char.IsDigit(r.Peek))
             {
-                StringBuilder result = new StringBuilder();
-
+                var result = new StringBuilder();
+                var foundDecimal = false;
                 while (!r.End)
                 {
                     if (char.IsDigit(r.Peek))
                     {
                         result.Append(r.Take());
                     }
+                    else if (!foundDecimal && r.Peek == '.')
+                    {
+                        result.Append(r.Take());
+                        foundDecimal = true;
+                    }
                     else
                     {
                         break;
                     }
                 }
 
-                return int.Parse(result.ToString(), CultureInfo.InvariantCulture);
+                if (!foundDecimal)
+                {
+                    return int.Parse(result.ToString(), CultureInfo.InvariantCulture); 
+                }
+                else
+                {
+                    return result.ToString(); // Leave as a string to support double, float, and decimal indicies
```

Comment:
I'd say if the target is a `IList` or `Array` then try to do the conversions there and then return `UnsetValue` if they fail.